{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ad523407-67d5-4810-adc6-c30b335fc354",
   "metadata": {},
   "source": [
    "# Karaka PN methylation basecalling.\n",
    "## Data & workflow\n",
    "- R10 chemistry\n",
    "- Sequenced on a minion in house.\n",
    "- from `final_summary_PAW31230_f467ca5d_9b2e8d74.txt`: protocol=sequencing/sequencing_MIN114_DNA_e8_2_400K:FLO-MIN114:SQK-LSK114:400"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ff245cb",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "DORADO=/workspace/hraijc/git_clones/dorado-1.0.2-linux-x64/bin/dorado\n",
    "WKDIR=/workspace/hraijc/TEMP/KarakaPN_ONT/\n",
    "READDIR=/workspace/hraijc/TEMP/KarakaPN_ONT/Karaka_PN/Karaka_PN/20260120_1200_MN19482_FBC78246_293559bf/pod5\n",
    "mkdir -p $WKDIR/log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "864ebac0-d0b5-4036-9f13-0d9362b022f5",
   "metadata": {},
   "source": [
    "# 00.Sup basecalling with 5mC. Skipping 6mA mods due to low quality calls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e34f977b-d75c-4a31-90d9-af984477f254",
   "metadata": {},
   "source": [
    "## Karaka Run 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f10f1ed-343d-4cb1-97cf-3f3a862c676c",
   "metadata": {},
   "outputs": [],
   "source": [
    "DORADO=/workspace/hraijc/git_clones/dorado-1.0.2-linux-x64/bin/dorado\n",
    "WKDIR=/workspace/hraijc/TEMP/KarakaPN_ONT/\n",
    "READDIR=/workspace/hraijc/TEMP/KarakaPN_ONT/Karaka_PN/Karaka_PN/20260120_1200_MN19482_FBC78246_293559bf/pod5\n",
    "mkdir -p $WKDIR/log\n",
    "\n",
    "LOG=$WKDIR/log\n",
    "OUT=$WKDIR/basecalling\n",
    "mkdir -p ${OUT}\n",
    "cd ${OUT}\n",
    "\n",
    "sbatch << EOF\n",
    "#!/bin/bash -e\n",
    "#SBATCH -J basecalling\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=8\n",
    "#SBATCH --mem=160G\n",
    "#SBATCH --output=${WKDIR}/log/%J.out\n",
    "#SBATCH --error=${WKDIR}/log/%J.err\n",
    "#SBATCH --time=47:00:00\n",
    "#SBATCH --gres=gpu:1\n",
    "\n",
    "cd $OUT\n",
    "\n",
    "$DORADO download --model dna_r10.4.1_e8.2_400bps_sup@v5.0.0\n",
    "$DORADO download --model dna_r10.4.1_e8.2_400bps_sup@v5.0.0_5mC_5hmC@v1\n",
    "\n",
    "$DORADO basecaller dna_r10.4.1_e8.2_400bps_sup@v5.0.0 \\\n",
    "    $READDIR \\\n",
    "     --modified-bases-models dna_r10.4.1_e8.2_400bps_sup@v5.0.0_5mC_5hmC@v1 \\\n",
    "    --min-qscore 15 > KarakaPN_ONT.bam\n",
    "\n",
    "\n",
    "EOF\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0ba4be2",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "DORADO=/workspace/hraijc/git_clones/dorado-1.0.2-linux-x64/bin/dorado\n",
    "WKDIR=/workspace/hraijc/TEMP/KarakaPN_ONT/\n",
    "READDIR=/workspace/hraijc/TEMP/KarakaPN_ONT/Karaka_PN_2/Karaka_PN_2/20260121_0937_MN19482_FBC78246_07078daa/pod5\n",
    "mkdir -p $WKDIR/log\n",
    "\n",
    "LOG=$WKDIR/log\n",
    "OUT=$WKDIR/basecalling\n",
    "mkdir -p ${OUT}\n",
    "cd ${OUT}\n",
    "\n",
    "sbatch << EOF\n",
    "#!/bin/bash -e\n",
    "#SBATCH -J basecalling\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=8\n",
    "#SBATCH --mem=160G\n",
    "#SBATCH --output=${WKDIR}/log/%J.out\n",
    "#SBATCH --error=${WKDIR}/log/%J.err\n",
    "#SBATCH --time=47:00:00\n",
    "#SBATCH --gres=gpu:1\n",
    "\n",
    "cd $OUT\n",
    "$DORADO basecaller dna_r10.4.1_e8.2_400bps_sup@v5.0.0 \\\n",
    "    $READDIR \\\n",
    "     --modified-bases-models dna_r10.4.1_e8.2_400bps_sup@v5.0.0_5mC_5hmC@v1 \\\n",
    "    --min-qscore 15 > KarakaPN_ONT2.bam\n",
    "\n",
    "\n",
    "\n",
    "EOF\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91b05514",
   "metadata": {},
   "source": [
    "## bam to fastq and qc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ee3634c",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "WKDIR=/workspace/hraijc/TEMP/KarakaPN_ONT/basecalling/\n",
    "LOG=$WKDIR/log\n",
    "mkdir -p $LOG\n",
    "cd $WKDIR\n",
    "\n",
    "\n",
    "sbatch << EOF\n",
    "#!/bin/bash\n",
    "#SBATCH -J nanocomp\n",
    "#SBATCH -o $LOG/%J.out \n",
    "#SBATCH -e $LOG/%J.err\n",
    "#SBATCH --cpus-per-task=2\n",
    "#SBATCH --mem=140G\n",
    "#SBATCH --time=14:00:00\n",
    "\n",
    "cd $WKDIR\n",
    "\n",
    "# load packages\n",
    "module load samtools/1.20\n",
    "module load htslib/1.20\n",
    "\n",
    "samtools view -@ 2 -O fastq KarakaPN_ONT.bam | bgzip -@ 2 > KarakaPN_ONT1.fastq.gz\n",
    "samtools view -@ 2 -O fastq KarakaPN_ONT2.bam | bgzip -@ 2 > KarakaPN_ONT2.fastq.gz\n",
    "cat KarakaPN_ONT1.fastq.gz KarakaPN_ONT2.fastq.gz > KarakaPN_ONT.fastq.gz\n",
    "\n",
    "module unload samtools/1.20\n",
    "module unload htslib/1.20\n",
    "\n",
    "# QC\n",
    "module load conda\n",
    "conda deactivate\n",
    "\n",
    "module load nanopack\n",
    "\n",
    "NanoPlot --fastq KarakaPN_ONT.fastq.gz --outdir nanoplot_out\n",
    "\n",
    "\n",
    "EOF\n",
    "\n",
    "#check outputs and delete scratch files. \n",
    "rm -rf KarakaPN_ONT1.fastq.gz\n",
    "rm -rf KarakaPN_ONT2.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a21f10fe",
   "metadata": {},
   "source": [
    "## Move to input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e18f8b3e-9d9f-4c0b-9fbe-dee2ebefafda",
   "metadata": {},
   "outputs": [],
   "source": [
    "mv /powerplant/workspace/hraijc/TEMP/KarakaPN_ONT/ /powerplant/workspace/hraijc/TEMP/KarakaPN_ONT_MinION_2026\n",
    "mv /powerplant/workspace/hraijc/TEMP/KarakaPN_ONT_MinION_2026 /powerplant/input/taonga/genomic/plant/Corynocarpus/laevigatus\n",
    "\n",
    "export odir=/input/taonga/genomic/plant/Corynocarpus/laevigatus/KarakaPN_ONT_MinION_2026\n",
    "find $odir -type d -exec chmod 755 {} \\; 2>/dev/null\n",
    "find $odir -type f -exec chmod 444 {} \\; 2>/dev/null\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
