{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "47d35bf0-8645-43c2-8788-a00c1152f85d",
   "metadata": {},
   "source": [
    "## Look at methylation levels using the Karaka v1 primary assembly:\n",
    "- Modified basecalling done by Ignacio Carvajal\n",
    "- Summarise global methylation levels\n",
    "- Generate percent methylaiton per 100k windows for Chen Wu's circos plot\n",
    "- Some of these scripts were adapted from code provided in this Repo: https://github.com/yusmiatiliau/Plant_skimseq_methylation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26f284a3-62ff-43b0-9858-4538d4ef1d1d",
   "metadata": {},
   "source": [
    "# 00.merge bams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88444262-c772-42e3-8de6-205b19b2bffc",
   "metadata": {},
   "outputs": [],
   "source": [
    "WKDIR=${BASE}/methylation\n",
    "OUT=${BASE}/methylation/00.merge_bams\n",
    "\n",
    "mkdir -p -v ${OUT}\n",
    "cd ${WKDIR}\n",
    "\n",
    "ml samtools/1.20\n",
    "\n",
    "sbatch << EOF\n",
    "#!/bin/bash\n",
    "#SBATCH -J merge\n",
    "#SBATCH --cpus-per-task=6\n",
    "#SBATCH --mem-per-cpu=104G\n",
    "#SBATCH --time=02:00:00\n",
    "#SBATCH -o $LOG/hrasrb_%J.out\n",
    "#SBATCH -e $LOG/hrasrb_%J.err\n",
    "#SBATCH --mail-user=Sarah.Bailey@plantandfood.co.nz\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "samtools merge -@ 6 -o ${OUT}/KarakaPN_ONT_merged.bam \\\n",
    "    ${IN}/KarakaPN_ONT.bam \\\n",
    "    ${IN}/KarakaPN_ONT2.bam\n",
    "EOF\n",
    "\n",
    "module unload samtools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fbf6dd7-c116-4d88-9a4f-1b47097e0ab0",
   "metadata": {},
   "source": [
    "# 01.align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5886533b-455c-4ffe-8c50-dd76ebff52e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: created directory '/powerplant/workspace/hrasrb/Karaka/methylation/01.align'\n",
      "Submitted batch job 9327202\n"
     ]
    }
   ],
   "source": [
    "WKDIR=${BASE}/methylation\n",
    "OUT=${BASE}/methylation/01.align\n",
    "REF=${REFERENCE_PATH}/karaka_v1.primary.chromosome-only.fa\n",
    "mkdir -p -v ${OUT}\n",
    "cd ${OUT}\n",
    "\n",
    "sbatch --dependency=afterok:9327201 << EOF\n",
    "#!/bin/bash\n",
    "#SBATCH -J dorado_align\n",
    "#SBATCH --cpus-per-task=6\n",
    "#SBATCH --mem-per-cpu=48G\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH -o $LOG/hrasrb_%J.out\n",
    "#SBATCH -e $LOG/hrasrb_%J.err\n",
    "#SBATCH --mail-user=Sarah.Bailey@plantandfood.co.nz\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "${MODULE_PATH}/dorado-0.9.6-linux-x64/bin/dorado aligner ${REF} \\\n",
    "    ${WKDIR}/00.merge_bams/KarakaPN_ONT_merged.bam --threads 6 > ${OUT}/karaka_v1.primary.5mC.bam\n",
    "\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68d7cbfd-f8cd-4ad0-b835-6f233e359f1c",
   "metadata": {},
   "source": [
    "# 02.sort and index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d48e9316-712b-40f1-94ec-a07a927946e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 9329686\n"
     ]
    }
   ],
   "source": [
    "WKDIR=${BASE}/methylation\n",
    "OUT=${BASE}/methylation/01.align\n",
    "\n",
    "mkdir -p -v ${OUT}\n",
    "cd ${OUT}\n",
    "\n",
    "ml samtools/1.20\n",
    "sbatch << EOF\n",
    "#!/bin/bash\n",
    "#SBATCH -J filt\n",
    "#SBATCH --cpus-per-task=4\n",
    "#SBATCH --mem-per-cpu=64G\n",
    "#SBATCH --time=01:00:00\n",
    "#SBATCH -o $LOG/hrasrb_%J.out\n",
    "#SBATCH -e $LOG/hrasrb_%J.err\n",
    "#SBATCH --mail-user=Sarah.Bailey@plantandfood.co.nz\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "# samtools sort -@ 2 ./karaka_v1.primary.5mC.bam -O BAM > ./karaka_v1.primary.5mC.sort.bam\n",
    "# samtools index -@ 2 ./karaka_v1.primary.5mC.sort.bam\n",
    "samtools view -F 260 -@4 -hb ./karaka_v1.primary.5mC.sort.bam > ./karaka_v1.primary.5mC.filt.bam\n",
    "samtools index -@4 ./karaka_v1.primary.5mC.filt.bam\n",
    "rm ./karaka_v1.primary.5mC.sort.bam\n",
    "EOF\n",
    "\n",
    "module unload samtools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dd2933f-10ee-4a9c-9707-49add0ca7ee6",
   "metadata": {},
   "source": [
    "# 03.check call accuracy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d2feb32f-eaf1-46b1-982a-5ebc1c6b3f7c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 9327326\n"
     ]
    }
   ],
   "source": [
    "WKDIR=${BASE}/methylation\n",
    "OUT=${BASE}/methylation/02.check_call_accuracy\n",
    "\n",
    "mkdir -p ${OUT}\n",
    "cd ${OUT}\n",
    "\n",
    "sbatch << EOF\n",
    "#!/bin/bash\n",
    "#SBATCH -J sample_probs\n",
    "#SBATCH --cpus-per-task=2\n",
    "#SBATCH --mem-per-cpu=16G\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH -o $LOG/hrasrb_%J.out\n",
    "#SBATCH -e $LOG/hrasrb_%J.err\n",
    "#SBATCH --mail-user=Sarah.Bailey@plantandfood.co.nz\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "${MODULE_PATH}/dist_modkit_v0.5.0_5120ef7/modkit sample-probs ${WKDIR}/01.align/karaka_v1.primary.5mC.sort.bam --hist --out-dir ${OUT}\n",
    "\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beb38aa0-e0d2-42e7-841d-fe13c72088eb",
   "metadata": {},
   "source": [
    "# 04.qualimap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e0855f4d-738f-4f34-b547-b7b04b18b7a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 9329688\n"
     ]
    }
   ],
   "source": [
    "ml apptainer/1.1\n",
    "\n",
    "WKDIR=${BASE}/methylation\n",
    "OUT=${BASE}/methylation/04.qualimap\n",
    "\n",
    "mkdir -p ${OUT}\n",
    "cd ${OUT}\n",
    "\n",
    "sbatch --dependency=afterok:9329686 << EOF\n",
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH -J qualimap\n",
    "#SBATCH -o ${LOG}/hrasrb_%J.out\n",
    "#SBATCH -e ${LOG}/hrasrb_%J.err\n",
    "#SBATCH --cpus-per-task=6\n",
    "#SBATCH --mem=80G\n",
    "#SBATCH --time=01:20:00\n",
    "#SBATCH --mail-user=Sarah.Bailey@plantandfood.co.nz\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "export TMPDIR=\"/workspace/\\$USER/tmp\"\n",
    "export APPTAINER_BINDPATH=\"\\$APPTAINER_BINDPATH,\\$TMPDIR:\\$TMPDIR,\\$TMPDIR:/tmp\"\n",
    "export SINGULARITY_CACHEDIR=\"/workspace/\\$USER/my_modules/singularity\"\n",
    "\n",
    "${MODULE_PATH}/singularity/qualimap_2.3--hdfd78af_0.sif qualimap bamqc \\\n",
    "    -bam ${WKDIR}/01.align/karaka_v1.primary.5mC.filt.bam \\\n",
    "    -nt 6 \\\n",
    "    -outdir ${OUT} \\\n",
    "    -outfile karaka_v1.primary.5mC.filt.html \\\n",
    "    -outformat HTML \\\n",
    "    --java-mem-size=80G \\\n",
    "    --paint-chromosome-limits\n",
    "\n",
    "EOF\n",
    "module unload apptainer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2e34b3d-9c26-40f9-8165-5ed59e7ada71",
   "metadata": {},
   "source": [
    "# 05.modkit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b55d26ec-c105-4406-ae96-2cec0c83f39a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 9329858\n"
     ]
    }
   ],
   "source": [
    "WKDIR=${BASE}/methylation\n",
    "OUT=${BASE}/methylation/05.modkit\n",
    "REF=/output/genomic/plant/Corynocarpus/laevigatus/genome/v1/assembly/karaka_v1.primary.chromosome-only.fa\n",
    "\n",
    "mkdir -p ${OUT}\n",
    "cd ${OUT}\n",
    "\n",
    "sbatch << EOF\n",
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH -J qualimap\n",
    "#SBATCH -o ${LOG}/hrasrb_%J.out\n",
    "#SBATCH -e ${LOG}/hrasrb_%J.err\n",
    "#SBATCH --cpus-per-task=4\n",
    "#SBATCH --mem=56G\n",
    "#SBATCH --time=01:00:00\n",
    "#SBATCH --mail-user=Sarah.Bailey@plantandfood.co.nz\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "${MODULE_PATH}/dist_modkit_v0.5.0_5120ef7/modkit pileup ${WKDIR}/01.align/karaka_v1.primary.5mC.filt.bam \"./karaka_v1.primary.5mC_t90_allChrs.pileup.bed\" --ref \"${REF}\" --motif CG 0 --motif CHG 0 --motif CHH 0 -t 4 --log-filepath ./t90.log --filter-threshold C:0.85 --mod-threshold m:0.90\n",
    "sed -i 's| |\\t|g' \"./karaka_v1.primary.5mC_t90_allChrs.pileup.bed\"\n",
    "\n",
    "EOF\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f69320c6-4371-4f48-b964-93f7e0e15465",
   "metadata": {},
   "source": [
    "## Run split_bed\n",
    "first split by methylation type, then by chr and then compress and index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "87a4f486-356b-4a3d-a023-6ad2494a0aee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 9331258\n"
     ]
    }
   ],
   "source": [
    "WKDIR=${BASE}/methylation\n",
    "OUT=${BASE}/methylation/05.modkit\n",
    "PREFIX=karaka_v1.primary.5mC_t90\n",
    "\n",
    "mkdir -p ${OUT}\n",
    "cd ${OUT}\n",
    "\n",
    "ml htslib\n",
    "sbatch << EOF\n",
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH -J compress_index\n",
    "#SBATCH -o ${LOG}/hrasrb_%J.out\n",
    "#SBATCH -e ${LOG}/hrasrb_%J.err\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=32G\n",
    "#SBATCH --time=01:00:00\n",
    "#SBATCH --mail-user=Sarah.Bailey@plantandfood.co.nz\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "# echo \"split unzipped bed files by context.\"\n",
    "# awk -v P=\"${PREFIX}\" '{fname=\\$4; gsub(/,/, \"_\", fname); print > (fname \"_\" P \"_allChrs.bed\")}' \"${PREFIX}_allChrs.pileup.bed\"\n",
    "# echo \"${PREFIX} done.\"\n",
    "\n",
    "# cat m_CG_0_${PREFIX}_allChrs.bed | \\\n",
    "#     awk -v S=\"${PREFIX}\" -v M=\"CG\" '{print > (\\$1 \"_\" S \"_\" M \".bed\")}'\n",
    "# echo \"Split CG context done\"\n",
    "# cat m_CHG_0_${PREFIX}_allChrs.bed | \\\n",
    "#     awk -v S=\"${PREFIX}\" -v M=\"CHG\" '{print > (\\$1 \"_\" S \"_\" M \".bed\")}'\n",
    "# echo \"Split CHG context done\"\n",
    "# cat m_CHH_0_${PREFIX}_allChrs.bed | \\\n",
    "#     awk -v S=\"${PREFIX}\" -v M=\"CHH\" '{print > (\\$1 \"_\" S \"_\" M \".bed\")}'\n",
    "# echo \"Split CHH context done\"\n",
    "\n",
    "for c in chr{1..23}; do\n",
    "    file=\"\\${c}_${PREFIX}_CG.bed\"\n",
    "    bgzip -f \"\\$file\"\n",
    "    tabix -f -p bed \"\\${file}.gz\"\n",
    "done\n",
    "echo \"Compress and index CG context done.\"\n",
    "for c in chr{1..23}; do\n",
    "    file=\"\\${c}_${PREFIX}_CHG.bed\"\n",
    "    bgzip -f \"\\$file\"\n",
    "    tabix -f -p bed \"\\${file}.gz\"\n",
    "done\n",
    "echo \"Compress and index CHG context done.\"\n",
    "for c in chr{1..23}; do\n",
    "    file=\"\\${c}_${PREFIX}_CHH.bed\"\n",
    "    bgzip -f \"\\$file\"\n",
    "    tabix -f -p bed \"\\${file}.gz\"\n",
    "done\n",
    "echo \"Compress and index CHH context done.\"\n",
    "\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0daefe0-cbf3-4e8c-b584-4774579c07b4",
   "metadata": {},
   "source": [
    "## Run filterMethylation_{min_cov}x.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "57babe09-9120-41e5-b079-06079fdbc4c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 9336313\n"
     ]
    }
   ],
   "source": [
    "WKDIR=${BASE}/methylation/06.filter\n",
    "\n",
    "cd ${WKDIR}\n",
    "\n",
    "ml pfr-python3\n",
    "\n",
    "sbatch << EOF\n",
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH -J plotMethylation\n",
    "#SBATCH -o ${LOG}/hrasrb_%J.out\n",
    "#SBATCH -e ${LOG}/hrasrb_%J.err\n",
    "#SBATCH --cpus-per-task=2\n",
    "#SBATCH --mem=16G\n",
    "#SBATCH --time=02:00:00\n",
    "#SBATCH --mail-user=Sarah.Bailey@plantandfood.co.nz\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "python3 ${REPO_PATH}/High-quality-genomes/Karaka/filterMethylation_5x.py\n",
    "\n",
    "EOF\n",
    "module unload pfr-python3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ebf90236-fc45-475c-8e88-6ab621706159",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 9343149\n"
     ]
    }
   ],
   "source": [
    "# combine chr files\n",
    "\n",
    "WKDIR=${BASE}/methylation/06.filter\n",
    "\n",
    "cd ${WKDIR}\n",
    "# cut -f 1 ${REFERENCE_PATH}/karaka_v1.primary.chromosome-only.fa.fai > ids.txt\n",
    "\n",
    "module load htslib/1.20\n",
    "\n",
    "# create and submit bash script\n",
    "sbatch << EOF\n",
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH -J create_allChrs # change\n",
    "#SBATCH --output=${LOG}/hrasrb_%j.out\n",
    "#SBATCH --error=${LOG}/hrasrb_%j.err\n",
    "#SBATCH --mail-user=Sarah.Bailey@plantandfood.co.nz\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --time=01:30:00 # Walltime\n",
    "#SBATCH --mem=32G\n",
    "\n",
    "while read -r chr; do\n",
    "    prefix=\"\\${chr}_karaka_v1.primary.5mC_t90_CHH_5x\"\n",
    "    bgzip -d -f \"${WKDIR}/\\${prefix}.bed.gz\"\n",
    "    cat \"${WKDIR}/\\${prefix}.bed\" >> karaka_v1.primary.5mC_t90_CHH.bed\n",
    "done < ids.txt\n",
    "echo \"combined chrs.\"\n",
    "\n",
    "sort -k1,1V -k2,2n karaka_v1.primary.5mC_t90_CHH.bed\n",
    "echo \"sorted.\"\n",
    "\n",
    "bgzip -f karaka_v1.primary.5mC_t90_CHH.bed\n",
    "echo \"compressed.\"\n",
    "\n",
    "tabix -f -p bed karaka_v1.primary.5mC_t90_CHH.bed.gz\n",
    "echo \"indexed.\"\n",
    "\n",
    "EOF\n",
    "module unload htslib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "33feac94-7623-44bf-b7ef-5c8ba50907c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: created directory '/powerplant/workspace/hrasrb/Karaka/methylation/07.modkit_stats'\n",
      "SBATCH_ARGS: --time 00:20:00 --mem 2G --cpus-per-task=2 --mail-user=Sarah.Bailey@plantandfood.co.nz --mail-type=ALL\n",
      "JOB_ARRAY_NAME: probs_array\n",
      "GROUP_SIZE: 1\n",
      "NUM_COMMANDS: 3\n",
      "Submitted batch job 9343163\n"
     ]
    }
   ],
   "source": [
    "WKDIR=${BASE}/methylation/07.modkit_stats\n",
    "TAG=karaka_v1\n",
    "mkdir -p -v ${WKDIR}\n",
    "cd ${WKDIR}\n",
    "\n",
    "## If file already exists delete it, otherwise text gets appended to existing file\n",
    "FILE=./bed.fofn\n",
    "if [ -f \"$FILE\" ] ; then\n",
    "    rm \"$FILE\"\n",
    "fi\n",
    "\n",
    "cat <<'EOF' >> ./bed.fofn\n",
    "${BASE}/methylation/07.modkit_stats/karaka_v1.primary.5mC_t90_CG.bed.gz\n",
    "${BASE}/methylation/07.modkit_stats/karaka_v1.primary.5mC_t90_CHG.bed.gz\n",
    "${BASE}/methylation/07.modkit_stats/karaka_v1.primary.5mC_t90_CHH.bed.gz\n",
    "EOF\n",
    "\n",
    "ml htslib\n",
    "cat $FILE | while read line\n",
    "do\n",
    "    PREFIX=$(basename ${line} .bed.gz)\n",
    "    echo \"${MODULE_PATH}/dist_modkit_v0.5.0_5120ef7/modkit stats --log-filepath \"${PREFIX}_stats.log\" --no-header --mod-codes \"m\" --regions ${IN}/32.circosPlot_paper/windows_chr.100k.txt -o \"${PREFIX}.stats.100k.bed\" ${line}\"\n",
    "done | abatch -j probs_array --time 00:20:00 --mem 2G --cpus-per-task=2 --mail-user=Sarah.Bailey@plantandfood.co.nz --mail-type=ALL | sbatch --dependency=afterok:9343149\n",
    "module unload htslib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22a1dd28-edb6-4331-be91-de33d53c411e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
