{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9a6806a8-c77a-4f1e-954b-14af107615e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set Variables\n",
    "WKDIR=/workspace/hraijc/HiC_trials/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5155c268-930d-4161-b13c-3c56f1cee009",
   "metadata": {},
   "outputs": [],
   "source": [
    "READ1=/workspace/hraijc/10p_hic_1.fastq.gz\n",
    "READ2=/workspace/hraijc/10p_hic_2.fastq.gz\n",
    "ASSEMBLY=/workspace/hraijc/priGClenfilt.fasta\n",
    "TEMP=/workspace/hraijc/TEMP/\n",
    "APREFIX=2n_pri"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "96c583b9-32d8-4a8f-93dd-c41bb3f0504a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd ${WKDIR}\n",
    "mkdir -p ${WKDIR}/log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8e3bce89-3e0e-4f9d-89ff-25afcec21b97",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sbatch << EOF\n",
    "#!/bin/bash\n",
    "#SBATCH -J hic\n",
    "#SBATCH -o ${LOGDIR}/%J.out\n",
    "#SBATCH -e ${LOGDIR}/%J.err\n",
    "#SBATCH --cpus-per-task=12\n",
    "#SBATCH --mem-per-cpu=16G\n",
    "#SBATCH --time=88:00:00\n",
    "\n",
    "module load bwa/0.7.18\n",
    "module load samtools/1.20\n",
    "\n",
    "\n",
    "cd ${WKDIR}\n",
    "\n",
    "### Align reads with BWA. Use -5SP for Hi-C reads.#############################\n",
    "#samtools faidx ${ASSEMBLY}\n",
    "#bwa index ${ASSEMBLY}\n",
    "#bwa mem -5SP -t12 ${ASSEMBLY} ${READ1} ${READ2} -o ${TEMP}/${APREFIX}.sam\n",
    "\n",
    "\n",
    "### Flag PCR Duplicates with SAMBLASTER #######################################\n",
    "/workspace/hraijc/git_clones/samblaster/samblaster -i ${TEMP}/${APREFIX}.sam -o ${TEMP}/${APREFIX}_marked_byread.sam\n",
    "\n",
    "\n",
    "### Remove unmmaped and non-primary aligned reads. Sort and Index bam files.###\n",
    "samtools view -S -b -h -@ 12 -F 2316 ${TEMP}/${APREFIX}_marked_byread.sam > ${TEMP}/${APREFIX}_presort_marked.bam\n",
    "samtools sort -@ 12 ${TEMP}/${APREFIX}_presort_marked.bam -o ${APREFIX}.bam\n",
    "\n",
    "### Run YAHS with contig and scaffolding error correction. ####################\n",
    "/workspace/hraijc/git_clones/yahs/yahs --no-contig-ec --no-mem-check --telo-motif CCCTAAA ${ASSEMBLY} ${APREFIX}.bam -o ${APREFIX}_yahs1\n",
    "\n",
    "EOF\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40ae61a6-db2d-4109-93ae-9b41828c0761",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Make heatmap of YAHS output ###############################################\n",
    "cd ${WKDIR}\n",
    "\n",
    "out=${APREFIX}_yahs1 # prefix of outfiles produced by YAHS.  \n",
    "contigs=${ASSEMBLY} # need to be indexed, i.e., test.fa.gz.fai in same directory\n",
    "wkdir=${WKDIR}\n",
    "\n",
    "\n",
    "sbatch --job-name=hic_mapyahs \\\n",
    "    -o ${WKDIR}/log/%J.out \\\n",
    "    -e ${WKDIR}/log/%J.err \\\n",
    "    --cpus-per-task=8 \\\n",
    "    --mem=24G \\\n",
    "    --time=03:10:00 \\\n",
    "    --export=out=$out,contigs=$contigs,wkdir=$wkdir \\\n",
    "    /workspace/hraijc/Gitrepos/High-quality-genomes/Methods/DNase_HiC/notebooks/yahs_contactmapgen2.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23ad400f-744f-47d2-b9e4-95b42688adda",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
